rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,145 @@
+fn main() {
+    let mut y: i32 = 50000;
+
+    while y > 0 {
+        y = y - 1;
+    }
+    assert!(y == 0);
+}
+```
+
+
+## User Experience
+
+Once this RFC has been stabilized, users will be able to synthesize loop contracts for loops and verify programs with the synthesized loop contracts with a new command line flag `--synthesize-loop-contracts`.
+
+### How the Synthesizer Works with Unwinding Numbers
+For each loop, users will be able to choose to unwind it for a certain number of times by specifying the unwinding numbers explicitly, or to synthesize loop contracts for it when no unwinding number specified. 
+
+We introduce a new way for users to specify unwinding numbers for a single loops rather than just for the whole functions: specifying loop numbers using labels. E.g., in the following program, both loops in the function `foo` will be unwound `5` times, while only the loop `inner` in the function `bar` will be unwound `2` times.
+```rust
+#[kani::unwind(5)]
+fn foo() {
+    let mut y: i32 = 50000;
+    let mut x: i32 = 2;
+    while y > 0 {
+        while x > 0 {
+            y = y - 1;
+            x = x - 1;
+        }
+        x = 2;
+    }
+    assert!(y == 0 || y == -1);
+}
+
+#[kani::unwind(inner:2)]
+fn bar() {
+    let mut y: i32 = 50000;
+    let mut x: i32 = 2;
+    `outer`: while y > 0 {
+        `inner`: while x > 0 {
+            y = y - 1;
+            x = x - 1;
+        }
+        x = 2;
+    }
+    assert!(y == 0 || y == -1);
+}
+```
+Then, we the flag `--synthesize-loop-contracts` is present, the synthesizer will synthesize loop contracts for only the loop `outer` in the function `bar` and unwind the other three loops.
+
+### Limit Resource Used by Synthesizer for Termination
+Without resource limit, an enumerative synthesizer may run forever to exhaust search space consist of infinite number of candidates, especially when there is no solution in the search space. So, for the guarantee of termination, we provide users options: <code>--limit-synthesis-time <ins>T</ins></code> to limit the running time of the synthesizer, and <code>--limit-synthesis-size <ins>N</ins></code> to limit the sizes (number of nodes in the AST of the candidate predicate) of loop-invariant candidates we search for during the enumeration.
+
+We encourage users to specify such limit while we hardcode some default limit at the same time. 
+
+### Output of Kani when Synthesizer Enable
+When the flag `--synthesize-loop-contracts` is provided, Kani will report different result for different cases
+1. When there exists some loop invariant in the candidate space with which all assertions can be proved, Kani will synthesize the loop contracts, verify the program with the synthesized loop contracts, and report verification SUCCESS;
+2. When there exists no loop invariants in the candidate space that can be used to prove all assertions within the resource limit, Kani will report the verification result with the best-effort-synthesized loop contracts.
```

Comment:
When no working candidate has been found in the search space within the specified limits